src/Azure.Functions.Cli/Helpers/AzureHelper.cs

```diff
@@ -357,9 +390,75 @@ private static async Task ArmHttpAsync(HttpMethod method, Uri uri, string access
             }
         }
 
-        public static async Task PrintFunctionsInfo(Site functionApp, string accessToken, string managementURL, bool showKeys)
+        private static HttpClient functionAppReadyClient = null;
+
+        internal static async Task WaitUntilFunctionAppReadyAsync(Site functionApp, string accessToken, string managementURL,
+            bool showKeys, HttpMessageHandler messageHandler = null)
         {
-            var functions = await GetFunctions(functionApp, accessToken, managementURL);
+            var masterKey = await GetMasterKeyAsync(functionApp.SiteId, accessToken, managementURL);
+
+            if (functionAppReadyClient == null)
+            {
+                HttpMessageHandler handler = messageHandler ?? new HttpClientHandler();
+                if (StaticSettings.IsDebug)
+                {
+                    handler = new LoggingHandler(handler);
+                }
+
+                functionAppReadyClient = new HttpClient(handler);
+                const string jsonContentType = "application/json";
+                functionAppReadyClient.DefaultRequestHeaders.Add("User-Agent", Constants.CliUserAgent);
+                functionAppReadyClient.DefaultRequestHeaders.Add("Accept", jsonContentType);
+            }
+
+            await RetryHelper.Retry(async () =>
+                  {
+                      functionAppReadyClient.DefaultRequestHeaders.Add("x-ms-request-id", Guid.NewGuid().ToString());
+                      var uri = new Uri($"https://{functionApp.HostName}/admin/host/status?code={masterKey}");
+                      var request = new HttpRequestMessage()
+                      {
+                          RequestUri = uri,
+                          Method = HttpMethod.Get
+                      };
+
+                      var response = await functionAppReadyClient.SendAsync(request);
+                      ColoredConsole.Write(".");
+
+                      if (response.IsSuccessStatusCode)
+                      {
+                          var json = await response.Content.ReadAsAsync<JToken>();
+                          var processUpTime = json["processUptime"]?.ToObject<int>() ?? 0;
+
+                          // Wait 60sec for the readiness of the worker process rebooting by deployment.
+                          if (processUpTime >= Constants.DefaultWorkerProcessUptimeReadiness)  
+                          {
+                              ColoredConsole.WriteLine(" done");
+                              return;
+                          }
+                          else
+                          {
+                              throw new Exception($"GetFunctions is not ready. Hostname: {functionApp.HostName}");
+                          }
+                      }
+                      else
+                      {
+                          throw new Exception($"Status check returns unsuccessful status. Hostname: {functionApp.HostName}");
```

Comment:
do not throw exceptions as this is being retried. Only throw exception if last retry is unsuccessfull

Reply:
Will approve, once this is addressed!

Reply:
Hi @pragnagopa 
Could you help me to clarify this?  I see the `Retry` functions looks retry by Exception. https://github.com/Azure/azure-functions-core-tools/blob/dev/src/Azure.Functions.Cli/Helpers/RetryHelper.cs#L23
I thought that we have two opportunity to retry. 1.  status is not good. (probably functions host is not started yet) and 2. functions host is running but worker is not ready. 

In this case, How can I detect the case 1 without Exception?

Reply:
Ahh thanks for clafication. Retryhelper expects an exception to be thrown and only displays if flag is set. This is OK!